### PR TITLE
Consistent ray.init return value 

### PR DESCRIPTION
**Pull request by ckw017**
## Why are these changes needed?

[Proposal document](https://docs.google.com/document/d/1ln7_fUST18GOz4jJnI_zN00hfczXY48V5Ajy6fCmJCE/edit#)

This PR changes the return value of `ray.init` when not in client mode to be a RayContext, which acts as a context manager and the same public fields as ClientContext , as well a disconnect method (calls shutdown under the hood).

To prevent breaking scripts that rely on accessing through dict methods, RayContext also subclasses collections.abc.Mapping (can be treated as an immutable dict). This behavior will be removed in 2.0, so deprecation warnings are raised when `__getitem__` is used. To make migration simple, an additional dict field `address_info` is added with the same values as the original return value.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ijrsvt — COMMENTED** [review]
Can we make a common Context subclass?

**Comment by mwtian** [comment]
Adding @iycheng who worked on client context and multi-client.

File: python/ray/worker.py [hunk 1]
```diff
@@ -577,6 +580,82 @@ def get_dashboard_url():
...
+    def __enter__(self):
+        pass
+
+    @abstractmethod
+    def __exit__(self):
+        pass
+
+
+@dataclass
+class RayContext(BaseContext, dict):
```
Comment (mwtian) [inline python/ray/worker.py:607]: Inherit from `collections.abc.MutableMapping` / `Mapping` seems more recommended.

**Review by mwtian — COMMENTED** [review]
Looks good. 1 minor comment.

**Review by ijrsvt — APPROVED** [review]
LGTM!

**Review by mwtian — APPROVED** [review]

File: python/ray/worker.py [hunk 2]
```diff
@@ -577,6 +581,84 @@ def get_dashboard_url():
...
+    Context manager for attached drivers.
+    """
+    dashboard_url: Optional[str]
+    python_version: str
+    ray_version: str
+    ray_commit: str
+    protocol_version = Optional[str]
+    address_info: Dict[str, Optional[str]]
+
+    def __init__(self, address_info: dict):
```
Comment (iycheng) [inline python/ray/worker.py:619]:
```
    def __init__(self, address_info: Dict[str, Optional[str]]):
```

File: python/ray/worker.py [hunk 3]
```diff
@@ -577,6 +581,84 @@ def get_dashboard_url():
...
+        self.dashboard_url = get_dashboard_url()
+        self.python_version = "{}.{}.{}".format(
+            sys.version_info[0], sys.version_info[1], sys.version_info[2])
+        self.ray_version = ray.__version__
+        self.ray_commit = ray.__commit__
+        # No client protocol version since this driver was intiialized
+        # directly
+        self.protocol_version = None
+        self.address_info = dict(address_info)
```
Comment (iycheng) [inline python/ray/worker.py:628]: why do we convert it again here?

**Comment by iycheng** [comment]
LGTM

File: python/ray/worker.py (hunk 3 above)
Comment (ckw017) [reply python/ray/worker.py]: That part wasn't needed

**Comment by AmeerHajAli** [comment]
@ckw017 , should we merge this?

**Comment by ckw017** [comment]
I pinged phillip for final approval, we can merge afterwards

**Comment by AmeerHajAli** [comment]
@ckw017 / @pcmoritz  any updates on this?

**Comment by ckw017** [comment]
@AmeerHajAli good to merge IMO, Philip seems busy but the implementation here is what was in his proposal document.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by AmeerHajAli** [comment]
@ckw017 , can you please rebase + format and I will merge this considering it has been 2+ weeks since we asked @pcmoritz  to review this?

**Pull request merged**